core/kernel/shared/sprintf.rb

```diff
@@ -0,0 +1,945 @@
+        format("%2$E", 0, 109.52).should == "1.095200E+02"
+        format("%2$f", 0, 10.952).should == "10.952000"
+        format("%2$g", 0, 12.1234).should == "12.1234"
+        format("%2$G", 0, 12.1234).should == "12.1234"
+        format("%2$a", 0, 196).should == "0x1.88p+7"
+        format("%2$A", 0, 196).should == "0X1.88P+7"
+
+        format("%2$c", 1, 97).should == "a"
+        format("%2$p", "a", []).should == "[]"
+        format("%2$s", "-", "abc").should == "abc"
+      end
+
+      it "raises exception if argument number is bigger than actual arguments list" do
+        -> () {
+          format("%4$d", 1, 2, 3)
+        }.should raise_error(ArgumentError)
+      end
+
+      it "ignores '-' sign" do
+        format("%2$d", 1, 2, 3).should == "2"
+        format("%-2$d", 1, 2, 3).should == "2"
+      end
+
+      it "raises ArgumentError exception when absolute and relative argument numbers are mixed" do
+        -> () {
+          format("%1$d %d", 1, 2)
+        }.should raise_error(ArgumentError)
+      end
+    end
+
+    describe "#" do
+      context "applies to format o" do
+        it "increases the precision until the first digit will be `0' if it is not formatted as complements" do
+          format("%#o", 87).should == "0127"
+        end
+
+        it "does nothing for negtive argument" do
+          format("%#o", -87).should == "..7651"
+        end
+      end
+
+      context "applies to formats bBxX" do
+        it "prefixes the result with 0x, 0X, 0b and 0B respectively for non-zero argument" do
+          format("%#b", 10).should == "0b1010"
+          format("%#b", -10).should == "0b..10110"
+          format("%#B", 10).should == "0B1010"
+          format("%#B", -10).should == "0B..10110"
+
+          format("%#x", 196).should == "0xc4"
+          format("%#x", -196).should == "0x..f3c"
+          format("%#X", 196).should == "0XC4"
+          format("%#X", -196).should == "0X..F3C"
+        end
+
+        it "does nothing for zero argument" do
+          format("%#b", 0).should == "0"
+          format("%#B", 0).should == "0"
+
+          format("%#o", 0).should == "0"
+
+          format("%#x", 0).should == "0"
+          format("%#X", 0).should == "0"
+        end
+      end
+
+      context "applies to formats aAeEfgG" do
+        it "forces a decimal point to be added, even if no digits follow" do
+          format("%#.0a", 16.25).should == "0x1.p+4"
+          format("%#.0A", 16.25).should == "0X1.P+4"
+
+          format("%#.0e", 100).should == "1.e+02"
+          format("%#.0E", 100).should == "1.E+02"
+
+          format("%#.0f", 123.4).should == "123."
+
+          format("%#g", 123456).should == "123456."
+          format("%#G", 123456).should == "123456."
+        end
+
+        # strange behavior
+        it "changes format from dd.dddd to exponential form for gG" do
+          format("%#.0g", 123.4).should_not == "123."
+          format("%#.0g", 123.4).should == "1.e+02"
```

Comment:
This one sounds weird indeed, could you open a bug and ask if it's intended?

Reply:
I'm not sure how that's weird. Without the `#`: `format("%.0g", 123.4) # => 1.e+02`. The doc says:
> For `g' and 'G', do not remove trailing zeros
That's exactly what happens here.
To see the actual effect of `#`:
```
sprintf("%.6g", 123.4) # => "123.4"
sprintf("%#.6g", 123.4) # => "123.400" 
```

Reply:
Ah, so the problem is 0 precision is specified in this case, so we just get exponential form:
```ruby
> "%.0g" % 123.4
=> "1e+02"
```
@andrykonchin Can you add a spec for that case and use @marcandre example for `#`?

Reply:
@eregon 
marcandre quoted already existed test cases

> Ah, so the problem is 0 precision is specified in this case, so we just get exponential form

It's still unclear to me because exponential form is used when default `dd.dddd` format looses precision. Adding `.` doesn't change anything and IMHO doesn't make sense for exponential form